File: src/main/java/com/mayonez/KeyInput.java
package com.mayonez;

import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;

public class KeyInput extends KeyAdapter {

	// Key Fields
	private boolean[] keys = new boolean[256];

	// KeyListener Methods

	@Override
	public void keyPressed(KeyEvent e) {
		int code = e.getKeyCode();
		keys[code] = true;
	}

	@Override
	public void keyReleased(KeyEvent e) {
		int code = e.getKeyCode();
		keys[code] = false;
	}

	// Getters and Setters

	public boolean keyDown(int keyCode) {
		return keys[keyCode];
	}

	public boolean keyDown(String keyName) {
		for (KeyMapping m : KeyMapping.values()) {
			if (m.toString().equalsIgnoreCase(keyName)) // if the desired mapping exists
				return keys[m.keyCode];
		}
		return false;
	}

	public int getAxis(String axisName) {
		for (KeyAxis a : KeyAxis.values()) {
			if (a.toString().equalsIgnoreCase(axisName))
				return a.value();
		}
		return 0;
	}

	/*
	 * Enum Definitions
	 */

	// TODO keyboard acceleration
	enum KeyAxis {

		VERTICAL(KeyMapping.DOWN, KeyMapping.UP), HORIZONTAL(KeyMapping.RIGHT, KeyMapping.LEFT);

		private int posKey, negKey;

		// make reference to key instead of code
		private KeyAxis(KeyMapping posKey, KeyMapping negKey) {
			this.posKey = posKey.keyCode();
			this.negKey = negKey.keyCode();
		}

//		// if contains key code
//		public boolean match(int keyCode) {
//			return keyCode == posKeyCode || keyCode == negKeyCode;
//		}

		public int value() {
			// "vector "method to make sure keys don't override each other
			int negComp = (Game.keyboard().keyDown(negKey)) ? -1 : 0;
			int posComp = (Game.keyboard().keyDown(posKey)) ? 1 : 0;
			return negComp + posComp;
		}

		@Override
		public String toString() {
			return name().toLowerCase();
		}

	}

	enum KeyMapping {

		// TODO read from file to assign keybinds
		UP(KeyEvent.VK_W), DOWN(KeyEvent.VK_S), LEFT(KeyEvent.VK_A), RIGHT(KeyEvent.VK_D), EXIT(KeyEvent.VK_ESCAPE),
		SPACE(KeyEvent.VK_SPACE);

		private int keyCode; // allow for multiple?

		private KeyMapping(int keyCode) {
			this.keyCode = keyCode;
			// KeyEvent.getKeyText(keyCode);
		}

		public int keyCode() {
			return keyCode;
		}

		@Override
		public String toString() {
			return name().toLowerCase();
		}

	}

}

File: src/main/kotlin/slavsquatsuperstar/mayonez/input/KeyInput.kt
package slavsquatsuperstar.mayonez.input

import org.lwjgl.glfw.GLFW.GLFW_PRESS
import org.lwjgl.glfw.GLFW.GLFW_RELEASE
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent

/**
 * The receiver for all keyboard-related input events.
 *
 * @author SlavSquatSuperstar
 */
// TODO game loop / callback methods shouldn't be accessible from API (called statically)
object KeyInput : KeyAdapter() {

    // Key Fields

    private val keysLast = BooleanArray(350) // continuously held
    private val keysDown = BooleanArray(350) // tapped once

    // Game Loop Methods

    @JvmStatic
    fun endFrame() {
        keysLast.fill(element = false)
    }

    // GLFW Callback Methods

    @SuppressWarnings("unused")
    @JvmStatic
    fun keyCallback(window: Long, key: Int, scancode: Int, action: Int, mods: Int) {
        if (action == GLFW_PRESS) {
            keysDown[key] = true
            keysLast[key] = true
        } else if (action == GLFW_RELEASE) {
            keysDown[key] = false
            keysLast[key] = false
        }
    }

    // KeyListener Overrides

    override fun keyPressed(e: KeyEvent) { // works in window begin frame, but not game update
        val code = e.keyCode
        if (!keyDown(code)) { // only register press if not down
            keysLast[code] = true
            keysDown[code] = true
        }
    }

    override fun keyReleased(e: KeyEvent) {
        val code = e.keyCode
        keysLast[code] = false
        keysDown[code] = false
    }

    // Keyboard Getters

    @JvmStatic
    fun keyDown(keyCode: Int): Boolean = keysDown[keyCode]

    @JvmStatic
    fun keyPressed(keyCode: Int): Boolean = keysLast[keyCode]

    /**
     * Returns whether any of the keys associated with the specified [KeyMapping] is continuously held down.
     *
     * @param keyName the name of the [KeyMapping]
     * @return if the specified key is down
     */
    @JvmStatic
    fun keyDown(keyName: String): Boolean {
        for (m in KeyMapping.values()) {
            if (m.name.equals(keyName, ignoreCase = true)) { // if the desired mapping exists
                for (code in m.keyCodes) if (keyDown(code)) return true
            }
        }
        return false
    }

    /**
     * Returns whether any of the keys associated with the specified [KeyMapping] has been pressed.
     *
     * @param keyName the name of the [KeyMapping]
     * @return if the specified key was pressed
     */
    /*
     * maybe too slow and not being called early enough
     */
    @JvmStatic
    fun keyPressed(keyName: String): Boolean {
        for (m in KeyMapping.values()) {
            if (m.name.equals(keyName, ignoreCase = true)) { // if the desired mapping exists
                for (code in m.keyCodes) if (keyPressed(code)) return true
            }
        }
        return false
    }

    @JvmStatic
    fun getAxis(axisName: String?): Int {
        for (a in KeyAxis.values())
            if (a.toString().equals(axisName, ignoreCase = true))
                return a.value()
        return 0
    }

}